File: candroid/candroid/src/main/java/com/ebuki/portal/fragment/PageDetailsFragment.java
/*
 * Copyright (C) 2016 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.instructure.candroid.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.View;

import com.instructure.candroid.R;
import com.instructure.candroid.delegate.Navigation;
import com.instructure.candroid.util.LockInfoHTMLHelper;
import com.instructure.candroid.util.Param;
import com.instructure.canvasapi2.StatusCallback;
import com.instructure.canvasapi2.managers.PageManager;
import com.instructure.canvasapi2.models.CanvasContext;
import com.instructure.canvasapi2.models.Page;
import com.instructure.canvasapi2.utils.ApiType;
import com.instructure.canvasapi2.utils.LinkHeaders;
import com.instructure.pandautils.utils.Const;
import com.instructure.pandautils.views.CanvasWebView;

import java.util.HashMap;
import java.util.Locale;

import retrofit2.Call;

public class PageDetailsFragment extends InternalWebviewFragment {

    // logic
    private String pageName;

    // asyncTasks
    private StatusCallback<Page> pageCallback;

    private Page page;

    @Override
    public FRAGMENT_PLACEMENT getFragmentPlacement(Context context) {return FRAGMENT_PLACEMENT.DETAIL; }

    @Override
    public String getFragmentTitle() {
        return getString(R.string.pages);
    }

    @Nullable
    @Override
    protected String getActionbarTitle() {
        return page != null ? page.getTitle() : null;
    }

    ///////////////////////////////////////////////////////////////////////////
    // LifeCycle
    ///////////////////////////////////////////////////////////////////////////


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setShouldLoadUrl(false);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setUpCallBack();

        if (pageName == null || pageName.equals(Page.FRONT_PAGE_NAME)) {
            PageManager.getFrontPage(getCanvasContext(), true, pageCallback);
        } else {
            PageManager.getPageDetails(getCanvasContext(), pageName, true, pageCallback);
        }
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        canvasWebView.setCanvasEmbeddedWebViewCallback(new CanvasWebView.CanvasEmbeddedWebViewCallback() {
            @Override
            public boolean shouldLaunchInternalWebViewFragment(String url) {
                return true;
            }

            @Override
            public void launchInternalWebViewFragment(String url) {
                InternalWebviewFragment.loadInternalWebView(getActivity(), (Navigation) getActivity(), InternalWebviewFragment.createBundle(getCanvasContext(), url, isLTITool()));
            }
        });
    }

    ///////////////////////////////////////////////////////////////////////////
    // Helpers
    ///////////////////////////////////////////////////////////////////////////

    @Override
    protected Page getModelObject() {
        return page;
    }

    @Override
    public HashMap<String, String> getParamForBookmark() {
        HashMap<String, String> map = getCanvasContextParams();
        if(Page.FRONT_PAGE_NAME.equals(pageName)) {
            map.put(Param.PAGE_ID, Page.FRONT_PAGE_NAME);
        } else if(!TextUtils.isEmpty(pageName)){
            map.put(Param.PAGE_ID, pageName);
        }
        return map;
    }

    @Override
    public boolean allowBookmarking() {
        return true;
    }

    ///////////////////////////////////////////////////////////////////////////
    // View
    ///////////////////////////////////////////////////////////////////////////

    ///////////////////////////////////////////////////////////////////////////
    // CallBack
    ///////////////////////////////////////////////////////////////////////////

    public void setUpCallBack(){
        pageCallback = new StatusCallback<Page>() {

            @Override
            public void onResponse(retrofit2.Response<Page> response, LinkHeaders linkHeaders, ApiType type) {
                if(!apiCheck()){
                    return;
                }

                getActivity().supportInvalidateOptionsMenu();
                page = response.body();

                if(page.getLockInfo() != null) {
                    String lockedMessage = LockInfoHTMLHelper.getLockedInfoHTML(page.getLockInfo(),getActivity(), R.string.lockedPageDesc, R.string.lockedAssignmentDescLine2);
                    populateWebView(lockedMessage, getString(R.string.pages));
                    return;
                }
                if (page.getBody() != null && !page.getBody().equals("null") && !page.getBody().equals("")) {
                    //this sets the width to be the device width and makes the images not be bigger than the width of the screen
                    populateWebView(page.getBody(), getString(R.string.pages));

                } else if (page.getBody() == null || page.getBody().endsWith("")) {
                    loadHtml("file:///android_asset/", getResources().getString(R.string.noPageFound), "text/html", "utf-8", null);
                }
                setupTitle(getActionbarTitle());
            }

            @Override
            public void onFail(Call<Page> response, Throwable error, int code) {
                if (response != null && code >= 400 && code < 500 && pageName != null && pageName.equals(Page.FRONT_PAGE_NAME)) {

                    String context;
                    if (getCanvasContext().getType() == CanvasContext.Type.COURSE) {
                        context = getString(R.string.course);
                    } else {
                        context = getString(R.string.group);
                    }

                    //We want a complete sentence.
                    context += ".";

                    //We want it to be lowercase.
                    context = context.toLowerCase(Locale.getDefault());

                    loadHtml("file:///android_asset/", getResources().getString(R.string.noPagesInContext) + " " + context, "text/html", "utf-8", null);
                }
            }
        };
    }

    ///////////////////////////////////////////////////////////////////////////
    // Intent
    ///////////////////////////////////////////////////////////////////////////

    @Override
    public void handleIntentExtras(Bundle extras) {
        super.handleIntentExtras(extras);

        if (getUrlParams() != null) {
            pageName = getUrlParams().get(Param.PAGE_ID);
        } else {
            pageName = extras.getString(Const.PAGE_NAME);
        }
    }

    public static Bundle createBundle(String pageName, CanvasContext canvasContext) {
        Bundle extras = createBundle(canvasContext);
        extras.putString(Const.PAGE_NAME, pageName);
        return extras;
    }
}

File: teacher/app/src/main/java/com/instructure/teacher/fragments/PageDetailsFragment.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.instructure.teacher.fragments

import android.graphics.Color
import android.os.Bundle
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.widget.Toast
import com.instructure.canvasapi2.models.CanvasContext
import com.instructure.canvasapi2.models.Course
import com.instructure.canvasapi2.models.Page
import com.instructure.canvasapi2.utils.APIHelper
import com.instructure.canvasapi2.utils.ApiPrefs
import com.instructure.pandautils.fragments.BasePresenterFragment
import com.instructure.pandautils.utils.*
import com.instructure.pandautils.views.CanvasWebView
import com.instructure.teacher.R
import com.instructure.teacher.activities.InternalWebViewActivity
import com.instructure.teacher.dialog.NoInternetConnectionDialog
import com.instructure.teacher.events.PageDeletedEvent
import com.instructure.teacher.events.PageUpdatedEvent
import com.instructure.teacher.factory.PageDetailsPresenterFactory
import com.instructure.teacher.interfaces.FullScreenInteractions
import com.instructure.teacher.interfaces.Identity
import com.instructure.teacher.interfaces.MasterDetailInteractions
import com.instructure.teacher.presenters.PageDetailsPresenter
import com.instructure.teacher.router.Route
import com.instructure.teacher.router.RouteMatcher
import com.instructure.teacher.utils.*
import com.instructure.teacher.viewinterface.PageDetailsView
import instructure.androidblueprint.PresenterFactory
import kotlinx.android.synthetic.main.fragment_page_details.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

class PageDetailsFragment : BasePresenterFragment<
        PageDetailsPresenter,
        PageDetailsView>(),
        PageDetailsView, Identity {

    private var mCanvasContext: CanvasContext by ParcelableArg(default = Course())
    private var mPage: Page by ParcelableArg(default = Page())
    private var mPageId: String by StringArg()

    override fun onHandleBackPressed(): Boolean {
        canvasWebView?.let {
            if (it.canGoBack()) {
                it.handleGoBack()
                return true
            }
        }
        return super.onHandleBackPressed()
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        super.onStop()
        EventBus.getDefault().unregister(this)
    }

    override fun onRefreshFinished() {
        loading.setGone()
    }

    override fun onRefreshStarted() {
        loading.setVisible()
    }

    override fun onReadySetGo(presenter: PageDetailsPresenter?) {

        if (mPage.isFrontPage) {
            presenter?.getFrontPage(mCanvasContext, true)
        } else if (!mPageId.isBlank()) {
            presenter?.getPage(mPageId, mCanvasContext, true)
        } else {
            presenter?.getPage(mPage.url ?: "", mCanvasContext, true)
        }
        setupToolbar()

        canvasWebView.canvasWebViewClientCallback = object : CanvasWebView.CanvasWebViewClientCallback {
            override fun openMediaFromWebView(mime: String, url: String, filename: String) {
                RouteMatcher.openMedia(activity, url)
            }

            override fun onPageFinishedCallback(webView: WebView, url: String) {
                loading.setGone()
            }

            override fun onPageStartedCallback(webView: WebView, url: String) {
                loading.setVisible()
            }

            override fun canRouteInternallyDelegate(url: String): Boolean = RouteMatcher.canRouteInternally(activity, url, ApiPrefs.domain, false)

            override fun routeInternallyCallback(url: String) {
                RouteMatcher.canRouteInternally(activity, url, ApiPrefs.domain, true)
            }
        }

        canvasWebView.webChromeClient = (object : WebChromeClient() {
            override fun onProgressChanged(view: WebView?, newProgress: Int) {
                super.onProgressChanged(view, newProgress)
                if (newProgress >= 100) {
                    loading?.setGone()
                }
            }
        })

        canvasWebView.canvasEmbeddedWebViewCallback = object : CanvasWebView.CanvasEmbeddedWebViewCallback {
            override fun launchInternalWebViewFragment(url: String) = activity.startActivity(InternalWebViewActivity.createIntent(activity, url, "", true))
            override fun shouldLaunchInternalWebViewFragment(url: String): Boolean = !RouteMatcher.canRouteInternally(activity, url, ApiPrefs.domain, false)

        }

        EventBus.getDefault().getStickyEvent(PageDeletedEvent::class.java)?.once(javaClass.simpleName + ".onResume()") {
            if (it.pageId == mPage.pageId) {
                if (activity is MasterDetailInteractions) {
                    (activity as MasterDetailInteractions).popFragment(mCanvasContext)
                } else if(activity is FullScreenInteractions) {
                    activity.finish()
                }
            }
        }
    }

    override fun getPresenterFactory(): PresenterFactory<PageDetailsPresenter> = PageDetailsPresenterFactory(mCanvasContext, mPage)
    override fun onPresenterPrepared(presenter: PageDetailsPresenter?) = Unit

    override fun layoutResId() = R.layout.fragment_page_details

    override val identity: Long? get() = mPage.pageId
    override val skipCheck: Boolean get() = false

    override fun populatePageDetails(page: Page) {
        mPage = page
        canvasWebView.loadHtml(page.body, page.title)
        setupToolbar()
    }

    override fun onError(stringId: Int) {
        Toast.makeText(context, stringId, Toast.LENGTH_SHORT).show()
    }

    private fun setupToolbar() {
        toolbar.setupMenu(R.menu.menu_page_details) { openEditPage(mPage) }

        toolbar.setupBackButtonWithExpandCollapseAndBack(this) {
            toolbar.updateToolbarExpandCollapseIcon(this)
            ViewStyler.themeToolbar(activity, toolbar, mCanvasContext.color, Color.WHITE)
            (activity as MasterDetailInteractions).toggleExpandCollapse()
        }

        toolbar.title = mPage.title
        if (!isTablet) {
            toolbar.subtitle = mCanvasContext.name
        }
        ViewStyler.themeToolbar(activity, toolbar, mCanvasContext.color, Color.WHITE)
    }

    private fun openEditPage(page: Page) {
        if(APIHelper.hasNetworkConnection()) {
            val args = CreateOrEditPageDetailsFragment.newInstanceEdit(mCanvasContext, page).arguments
            RouteMatcher.route(context, Route(CreateOrEditPageDetailsFragment::class.java, mCanvasContext, args))
        } else {
            NoInternetConnectionDialog.show(fragmentManager)
        }
    }

    @Suppress("unused")
    @Subscribe(threadMode = ThreadMode.MAIN, sticky = true)
    fun onPageUpdated(event: PageUpdatedEvent) {
        event.once(javaClass.simpleName) {
            // need to set a flag here. Because we use the event bus in the fragment instead of the presenter for unit testing purposes,
            // when we come back to this fragment it will go through the life cycle events again and the cached data will immediately
            // overwrite the data from the network if we refresh the presenter from here.
            event.get { mPage = it }
        }
    }

    companion object {
        const val PAGE = "pageDetailsPage"

        const val PAGE_ID = "pageDetailsId"

        @JvmStatic
        fun makeBundle(page: Page): Bundle = Bundle().apply { putParcelable(PageDetailsFragment.PAGE, page) }

        @JvmStatic
        fun makeBundle(pageId: String): Bundle = Bundle().apply { putString(PageDetailsFragment.PAGE_ID, pageId) }


        @JvmStatic
        fun newInstance(canvasContext: CanvasContext, args: Bundle) = PageDetailsFragment().apply {
            with(args) {
                if (containsKey(PAGE)) {
                    mPage = getParcelable(PAGE)
                }
                if (containsKey(PAGE_ID)) {
                    mPageId = getString(PAGE_ID)
                }
            }

            mCanvasContext = canvasContext
        }
    }
}
